### PR TITLE
Move KafkaCruiseControl to depend on cruise-control-core

### PR DESCRIPTION
**Pull request by becketqin**
This patch changes KafkaCruiseControl to depend on cruise-control-core.

### CODE REVIEW

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Host.java [hunk 1]
```diff
@@ -161,7 +157,7 @@ void addReplica(Replica replica) {
   Replica removeReplica(int brokerId, TopicPartition tp) {
     Broker broker = _brokers.get(brokerId);
     if (broker == null) {
-      throw new IllegalStateException(String.format("Cannot remove replica for %s from broker %s because "
+      throw new IllegalStateException(String.format("Cannot remove replica for %s from broker broker %s because "
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Host.java]: repeated word: `broker` x 2

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Load.java [hunk 2]
```diff
@@ -102,50 +71,36 @@ public static int maxNumSnapshots() {
...
    */
   public double expectedUtilizationFor(Resource resource) {
-    if (_snapshotsByTime.isEmpty()) {
+    if (_metricValues.isEmpty()) {
       return 0.0;
     }
-    return resource == Resource.DISK ? Math.max(0, _snapshotsByTime.get(0).utilizationFor(resource)) :
-        Math.max(0, _accumulatedUtilization[resource.id()] / _snapshotsByTime.size());
+    MetricValues metricValues = _metricValues.valuesFor(KafkaCruiseControlMetricDef.resourceToMetricId(resource));
+    return  resource.equals(Resource.DISK) ? metricValues.latest() : metricValues.avg();
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Load.java]: Extra space after return.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Load.java [hunk 3]
```diff
@@ -154,59 +109,35 @@ void setLoadFor(Resource resource, Map<Long, Double> loadBySnapshotTime) throws
...
-    if (!_snapshotsByTime.isEmpty() && snapshot.time() >= _snapshotsByTime.get(_snapshotsByTime.size() - 1).time()) {
-      throw new ModelInputException("Attempt to push an out of order snapshot with timestamp " + snapshot.time() +
-                                        " to a replica. Existing snapshot times: " +
-                                        Arrays.toString(allSnapshotTimes()));
-    }
-    _snapshotsByTime.add(snapshot);
-    for (Resource r : Resource.cachedValues()) {
-      double utilization = _accumulatedUtilization[r.id()];
-      _accumulatedUtilization[r.id()] = utilization + snapshot.utilizationFor(r);
+  void initializeMetricValues(AggregatedMetricValues aggregatedMetricValues, List<Long> windows) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/model/Load.java]: Missing JavaDoc for parameter `windows`.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java [hunk 4]
```diff
@@ -183,36 +191,25 @@ public LoadMonitorState state(OperationProgress operationProgress) {
     double minMonitoredPartitionsPercentage = _defaultModelCompletenessRequirements.minMonitoredPartitionsPercentage();
 
     // Get the window to monitored partitions percentage mapping.
-    Map<Long, Double> windowToMonitoredPercentage =
-        _metricCompletenessChecker.monitoredPercentages(clusterModelGeneration(),
-                                                        clusterAndGeneration.cluster(),
-                                                        totalNumPartitions);
+    SortedMap<Long, Float> partitionCoverageByWindows =
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java]: Reminder for the future name change: we removed the term `coverage` from our previous patch.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java [hunk 5]
```diff
@@ -663,6 +545,19 @@ private String getRackHandleNull(Node node) {
     return brokersWithPartitions;
   }
 
+  private Map<TopicPartition, List<SampleFlaw>> partitionSampleFlaws(Map<PartitionEntity, ValuesAndExtrapolations> valuesAndExtrapolations) {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitor.java]: Reminder for the future name change patch: we removed the term `flaw` from our previous patch.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsReporterSampler.java [hunk 6]
```diff
@@ -52,7 +53,8 @@ public Samples getSamples(Cluster cluster,
                             Set<TopicPartition> assignedPartitions,
                             long startTimeMs,
                             long endTimeMs,
-                            SamplingMode mode) throws MetricSamplingException {
+                            SamplingMode mode,
+                            MetricDef metricDef) throws MetricSamplingException {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsReporterSampler.java]: Does the exception `MetricSamplingException` ever thrown? Can we remove it and the corresponding import?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/PartitionMetricSample.java [hunk 7]
```diff
@@ -4,109 +4,32 @@
...
 import com.linkedin.kafka.cruisecontrol.metricsreporter.exception.UnknownVersionException;
+import com.linkedin.kafka.cruisecontrol.monitor.metricdefinition.KafkaCruiseControlMetricDef;
 import java.nio.ByteBuffer;
 import org.apache.kafka.common.TopicPartition;
 
 import java.util.Date;
-import java.util.HashMap;
 import java.util.Map;
 
+import static com.linkedin.kafka.cruisecontrol.monitor.metricdefinition.KafkaCruiseControlMetricDef.*;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/PartitionMetricSample.java]: Should we use the full import path rather than the wildcard?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * This class aggregates the metric samples generated by the MetricFetcher.
+ * <p>
+ * The metric sample aggregator performs the sanity check on the samples and aggregate the samples into the
+ * corresponding window. Because the Kafka broker does not report the bytes in rate for the followers,
+ * we assume the sample we get are only from the leaders, and we are going to derive the follower metrics based on
+ * the leader metrics.
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java]: Should we note that this behavior will change with Kafka 0.11?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+   */
+  public MetricSampleAggregationResult<String, PartitionEntity> aggregate(MetadataClient.ClusterAndGeneration clusterAndGeneration,
+                                                                          long from,
+                                                                          long to,
+                                                                          ModelCompletenessRequirements requirements,
+                                                                          OperationProgress operationProgress)
+      throws NotEnoughValidWindowsException {
+    RetrievingMetrics step = new RetrievingMetrics();
+    try {
+      operationProgress.addStep(step);
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java]: Should we check for the case where `operationProgress` could be `null`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+                                 true);
+    MetricSampleCompleteness<String, PartitionEntity> completeness = completeness(-1, Long.MAX_VALUE, options);
+    return windowIndexesToWindows(completeness.validWindowIndexes(), _windowMs);
+  }
+
+  /**
+   * Get the valid partitions percentage across all the windows.
+   *
+   * @param clusterAndGeneration the current cluster and generation.
+   * @return The percentage of valid partitions acorss all the windows.
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java]: `acorss` --> `across`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  public SampleFlaw(long snapshotWindow, Extrapolation imputation) {
+    _snapshotWindow = snapshotWindow;
+    _imputation = imputation;
+  }
+
+  public long snapshotWindow() {
+    return _snapshotWindow;
+  }
+
+  public Extrapolation imputation() {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java]: `imputation `--> `extrapolation`

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+package com.linkedin.kafka.cruisecontrol.monitor.sampling.aggregator;
+
+import com.linkedin.cruisecontrol.monitor.sampling.aggregator.Extrapolation;
+
+
+/**
+ * The sample flaw for a partition that is still treated as valid.
+ */
+public class SampleFlaw {
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java]: Public methods of this class are missing JavaDoc.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  private final long _snapshotWindow;
+  private final Extrapolation _imputation;
+
+  public SampleFlaw(long snapshotWindow, Extrapolation imputation) {
+    _snapshotWindow = snapshotWindow;
```
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java]: `_snapshotWindow` --> `_window`?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java (hunk 12 above)
Comment (efeg) [inline cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/SampleFlaw.java]: I believe we discussed to replace the `flaw` keyword with a word that says `withExtrapolation`?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java [hunk 14]
```diff
@@ -88,68 +94,67 @@ public void testStateWithOnlyActiveSnapshotWindow() {
...
     // populate the metrics aggregator.
-    // two samples for each partition except T1P1
-    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, T0P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, T0P1, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, T1P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(2, 1, aggregator, T1P1, 0, SNAPSHOT_WINDOW_MS);
+    // two samples for each partition except T1P1. It has no sample in the first window and one in the second window.
+    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, PE_T0P0, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, PE_T0P1, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(2, 4, aggregator, PE_T1P0, 0, WINDOW_MS, METRIC_DEF);
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Did we previously add the following line by mistake?: 
`CruiseControlUnitTestUtils.populateSampleAggregator(2, 1, aggregator, T1P1, 0, SNAPSHOT_WINDOW_MS);`

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java [hunk 15]
```diff
@@ -88,68 +94,67 @@ public void testStateWithOnlyActiveSnapshotWindow() {
...
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T0P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T0P1, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T1P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 1, aggregator, T1P1, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(1, 1, aggregator, T1P1, 0, SNAPSHOT_WINDOW_MS);
+    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, PE_T0P0, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, PE_T0P1, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, PE_T1P0, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(1, 2, aggregator, PE_T1P1, 0, WINDOW_MS, METRIC_DEF);
+    CruiseControlUnitTestUtils.populateSampleAggregator(1, 2, aggregator, PE_T1P1, 2, WINDOW_MS, METRIC_DEF);
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Are the changes in this and previous line intentional?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java [hunk 16]
```diff
@@ -88,68 +94,67 @@ public void testStateWithOnlyActiveSnapshotWindow() {
...
     assertEquals(2, state.monitoredSnapshotWindows().size());
-    assertEquals(1.0, state.monitoredSnapshotWindows().get(SNAPSHOT_WINDOW_MS), 0.0);
-    assertEquals(0.5, state.monitoredSnapshotWindows().get(SNAPSHOT_WINDOW_MS * 2), 0.0);
+    assertEquals(1.0, state.monitoredSnapshotWindows().get(WINDOW_MS), 0.0);
+    assertEquals(0.5, state.monitoredSnapshotWindows().get(WINDOW_MS * 2), 0.0);
 
 
     // Back fill a sample for T1P1
-    CruiseControlUnitTestUtils.populateSampleAggregator(1, 1, aggregator, T1P1, 1, SNAPSHOT_WINDOW_MS);
+    CruiseControlUnitTestUtils.populateSampleAggregator(1, 3, aggregator, PE_T1P1, 1, WINDOW_MS, METRIC_DEF);
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Is this change intentional?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java [hunk 17]
```diff
@@ -281,22 +282,21 @@ public void testClusterWithInvalidPartitions()
 
     // populate the metrics aggregator.
     // two samples for each partition except T1P1
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T0P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T0P1, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 4, aggregator, T1P0, 0, SNAPSHOT_WINDOW_MS);
-    CruiseControlUnitTestUtils.populateSampleAggregator(3, 1, aggregator, T1P1, 0, SNAPSHOT_WINDOW_MS);
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Did we intentionally remove this line?

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregatorTest.java [hunk 18]
```diff
@@ -0,0 +1,616 @@
...
+import org.junit.Test;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import static com.linkedin.kafka.cruisecontrol.monitor.metricdefinition.KafkaCruiseControlMetricDef.*;
+import static org.junit.Assert.*;
```
Comment (efeg) [inline cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregatorTest.java]: Should we use wildcard or specify what we import (this and prev. line)?

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/CapacityGoal.java [hunk 19]
```diff
@@ -195,7 +194,7 @@ private void ensureUtilizationUnderCapacity(ClusterModel clusterModel) throws Op
         }
       }
       // Broker-level violation check.
-      if (resource.isBrokerResource()) {
+      if (!resource.isHostResource()) {
```
[the remark this answers is not in the document]
Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/CapacityGoal.java]: This was not intentional. Will change it back.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/metricdefinition/KafkaCruiseControlMetricDef.java [hunk 20]
```diff
@@ -0,0 +1,113 @@
...
+import java.util.Map;
+
+import static com.linkedin.cruisecontrol.metricdef.ValueComputingStrategy.AVG;
+import static com.linkedin.cruisecontrol.metricdef.ValueComputingStrategy.LATEST;
+
+
+/**
+ * The metric definitions of Kafka Cruise Control.
+ *
+ * The class maps the raw metric types to metric definitions to be used by the analyzer and linear regression
```
[the remark this answers is not in the document]
Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/metricdefinition/KafkaCruiseControlMetricDef.java]: As of now the linear regression model is not an official feature yet. We should probably avoid talk about this too much. I'll just say it is still in development.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsReporterSampler.java (hunk 6 above)
Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/CruiseControlMetricsReporterSampler.java]: This method inherits the interface. Other implementations may throw this exception.

File: cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java (hunk 9 above)
Comment (becketqin) [reply cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java]: Our check is indeed a little loose here primarily because we control the input as well. We know that the `clusterGeneration`, `requirements` and `operationProgress` will not be null.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java (hunk 14 above)
Comment (becketqin) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: The extrapolation behavior changed a little. So the conditions need to be adjusted accordingly. More specifically, previously if there are less than 1/2 of required samples in a window and `AVG_ADJACENT` does not work, it is considered as `NO_VALID_EXTRAPOLATION`. Now it is `FORCED_INSUFFICIENT`.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java (hunk 15 above)
Comment (becketqin) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Yes, because of the reasons explained above.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java (hunk 16 above)
Comment (becketqin) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Yes, for the same reasons above.

File: cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java (hunk 17 above)
Comment (becketqin) [reply cruise-control/src/test/java/com/linkedin/kafka/cruisecontrol/monitor/LoadMonitorTest.java]: Yes, for the same reasons above.

**Comment by efeg** [comment]
LGTM -- after fixing the style error so that circleci is happy.
